File: cpp/src/bubble_sort.cpp
#include "algorithms.h"
#include <thread>
#include <chrono>
#include <iostream>

// Bubble Sort with visualization
// Time Complexity: O(n^2) - quadratic
// Space Complexity: O(1) - constant
//
// How it works:
// 1. Compare adjacent elements
// 2. Swap if they're in wrong order
// 3. Repeat until no more swaps needed
void bubbleSort(std::vector<int>& array, Visualizer& viz) {
    int n = array.size();
    std::vector<bool> sorted(n, false);
    int totalComparisons = 0;
    int totalSwaps = 0;
    auto startTime = std::chrono::high_resolution_clock::now();

    std::cout << "\n========================================\n";
    std::cout << "Starting Bubble Sort\n";
    std::cout << "========================================\n";
    std::cout << "Array size: " << n << "\n";
    std::cout << "Worst case: O(n^2) = " << (n * n) << " comparisons\n";
    std::cout << "Best case: O(n) = " << n << " comparisons\n";
    std::cout << "========================================\n\n";

    for (int i = 0; i < n - 1; i++) {
        bool swapped = false;
        int passComparisons = 0;
        int passSwaps = 0;

        std::cout << "Pass " << (i + 1) << "/" << (n - 1) << "... ";

        for (int j = 0; j < n - i - 1; j++) {
            totalComparisons++;
            passComparisons++;
            // Check for quit
            if (viz.shouldQuit()) return;

            // Visualize comparison
            viz.draw(array, j, j + 1, sorted);
            viz.playTone(array[j]);

            // The actual bubble sort logic
            if (array[j] > array[j + 1]) {
                std::swap(array[j], array[j + 1]);
                swapped = true;
                totalSwaps++;
                passSwaps++;
            }

            // Delay so we can see the visualization
            std::this_thread::sleep_for(std::chrono::milliseconds(viz.getDelayMs()));
        }

        // Mark the last element of this pass as sorted
        sorted[n - i - 1] = true;

        std::cout << passComparisons << " comparisons, " << passSwaps << " swaps\n";

        // If no swaps occurred, the array is sorted
        if (!swapped) {
            std::cout << "\nArray is sorted! Early termination at pass " << (i + 1) << "\n";
            // Mark all remaining elements as sorted
            for (int k = 0; k < n - i - 1; k++) {
                sorted[k] = true;
            }
            break;
        }
    }

    auto endTime = std::chrono::high_resolution_clock::now();
    auto duration = std::chrono::duration_cast<std::chrono::milliseconds>(endTime - startTime);

    std::cout << "\n========================================\n";
    std::cout << "Bubble Sort Complete!\n";
    std::cout << "========================================\n";
    std::cout << "Total comparisons: " << totalComparisons << "\n";
    std::cout << "Total swaps: " << totalSwaps << "\n";
    std::cout << "Time elapsed: " << duration.count() << "ms\n";
    std::cout << "Time complexity: O(n^2)\n";
    std::cout << "Space complexity: O(1)\n";
    std::cout << "========================================\n";

    // Final visualization showing all bars in green
    viz.draw(array, -1, -1, sorted);
    std::this_thread::sleep_for(std::chrono::milliseconds(1000));
}

File: rust/src/algorithms/bubble_sort.rs
use macroquad::prelude::*;
use std::thread;
use std::time::{Duration, Instant};

use crate::algorithms::SortAlgorithm;
use crate::visualizer::{AudioSystem, draw_array};

// Bubble Sort with visualization
// Time Complexity: O(n^2) - quadratic
// Space Complexity: O(1) - constant
//
// How it works:
// 1. Compare adjacent elements
// 2. Swap if they're in wrong order
// 3. Repeat until no more swaps needed
pub async fn bubble_sort(array: &mut [usize], algorithm: SortAlgorithm, audio: &AudioSystem, array_size: usize, delay_ms: u64) {
    let n = array.len();
    let mut sorted = vec![false; n];
    let mut total_comparisons = 0;
    let mut total_swaps = 0;
    let start_time = Instant::now();

    println!("\n========================================");
    println!("Starting Bubble Sort");
    println!("========================================");
    println!("Array size: {}", n);
    println!("Worst case: O(n^2) = {} comparisons", n * n);
    println!("Best case: O(n) = {} comparisons", n);
    println!("========================================\n");

    for i in 0..n - 1 {
        let mut swapped = false;
        let mut pass_comparisons = 0;
        let mut pass_swaps = 0;

        print!("Pass {}/{}... ", i + 1, n - 1);

        for j in 0..n - i - 1 {
            total_comparisons += 1;
            pass_comparisons += 1;

            // Visualize comparison
            draw_array(array, Some(j), Some(j + 1), &sorted, algorithm, array_size, delay_ms);
            next_frame().await;

            // Play tone for current element
            audio.play_tone(array[j], array_size);

            // The actual bubble sort logic
            if array[j] > array[j + 1] {
                array.swap(j, j + 1);
                swapped = true;
                total_swaps += 1;
                pass_swaps += 1;
            }

            // Delay so we can see the visualization
            thread::sleep(Duration::from_millis(delay_ms));
        }

        // Mark the last element of this pass as sorted
        sorted[n - i - 1] = true;

        println!("{} comparisons, {} swaps", pass_comparisons, pass_swaps);

        // If no swaps occurred, the array is sorted
        if !swapped {
            println!("\nArray is sorted! Early termination at pass {}", i + 1);
            // Mark all remaining elements as sorted
            for item in sorted.iter_mut().take(n - i - 1) {
                *item = true;
            }
            break;
        }
    }

    let duration = start_time.elapsed();

    println!("\n========================================");
    println!("Bubble Sort Complete!");
    println!("========================================");
    println!("Total comparisons: {}", total_comparisons);
    println!("Total swaps: {}", total_swaps);
    println!("Time elapsed: {}ms", duration.as_millis());
    println!("Time complexity: O(n^2)");
    println!("Space complexity: O(1)");
    println!("========================================");

    // Final visualization showing all bars in green
    draw_array(array, None, None, &sorted, algorithm, array_size, delay_ms);
    next_frame().await;
    thread::sleep(Duration::from_millis(1000));
}
